afanasy/src/watch/listtasks.cpp

```diff
@@ -123,157 +124,88 @@ void ListTasks::contextMenuEvent(QContextMenuEvent *event)
-						connect( actionid, SIGNAL( triggeredId( int ) ), this, SLOT( actTaskStdOut( int ) ));
-						submenu->addAction( actionid);
-					}
-					menu.addMenu( submenu);
-				}
-			}
-
-			action = new QAction( "Log", this);
-			connect( action, SIGNAL( triggered() ), this, SLOT( actTaskLog() ));
-			menu.addAction( action);
-
-			action = new QAction( "Info", this);
-			connect( action, SIGNAL( triggered() ), this, SLOT( actTaskInfo() ));
-			menu.addAction( action);
-
-			action = new QAction( "Listen", this);
-			connect( action, SIGNAL( triggered() ), this, SLOT( actTaskListen() ));
-			menu.addAction( action);
-
-			action = new QAction( "Error Hosts", this);
-			connect( action, SIGNAL( triggered() ), this, SLOT( actTaskErrorHosts() ));
-			menu.addAction( action);
-
-			std::vector<std::string> files = ((ItemJobTask*)(item))->genFiles();
-			if( files.size())
-			{
-				if( af::Environment::getPreviewCmds().size() > 0 )
-				{
-					menu.addSeparator();
-
-					action = new QAction( "Browse Files...", this);
-					connect( action, SIGNAL( triggered() ), this, SLOT( actBrowseFolder() ));
-					menu.addAction( action);
-
-				if( ((ItemJobTask*)(item))->isBlockNumeric() )
-				{
-					QMenu * submenu_cmd = new QMenu( "Preview", this);
-					int p = 0;
-					for( std::vector<std::string>::const_iterator it = af::Environment::getPreviewCmds().begin(); it != af::Environment::getPreviewCmds().end(); it++, p++)
-					{
-						if( files.size() > 1)
-						{
-							QString file = afqt::stoq((*it).c_str());
-							QMenu * submenu_img = new QMenu( QString("%1").arg( file), this);
-							for( int i = 0; i < files.size(); i++)
-							{
-								QString imgname = file.right(99);
-								ActionIdId * actionid = new ActionIdId( p, i, imgname, this);
-								connect( actionid, SIGNAL( triggeredId(int,int) ), this, SLOT( actTaskPreview(int,int) ));
-								submenu_img->addAction( actionid);
-							}
-							submenu_cmd->addMenu( submenu_img);
-						}
-						else
-						{
-							ActionIdId * actionid = new ActionIdId( p, 0, QString("%1").arg( QString::fromUtf8((*it).c_str())), this);
-							connect( actionid, SIGNAL( triggeredId(int,int) ), this, SLOT( actTaskPreview(int,int) ));
-							submenu_cmd->addAction( actionid);
-						}
-					}
-					menu.addMenu( submenu_cmd);
-				}
-				}
-			}
+			ItemJobTask *itemTask = static_cast<ItemJobTask*>(item);
+			
+			// Operations on the current task item
+			itemTask->generateMenu(o_menu);
 
-			menu.addSeparator();
+			o_menu.addSeparator();
+			// Operations on all the selected tasks
 
 			action = new QAction( "Skip Tasks", this);
 			connect( action, SIGNAL( triggered() ), this, SLOT( actTasksSkip() ));
-			menu.addAction( action);
+			o_menu.addAction( action);
 
 			action = new QAction( "Restart Tasks", this);
 			connect( action, SIGNAL( triggered() ), this, SLOT( actTasksRestart() ));
-			menu.addAction( action);
+			o_menu.addAction( action);
 
```

Comment:
I had to keep these two actions here because they can be called on a whole selection of tasks. I would have liked to put the logic in `ItemJobTask` anyway, but did of find a clean way to handle the multiplicity of items.